File: adapters/go/backfill/backfill.go
// Package backfill runs a workflow over many run_ids with bounded concurrency
// and aggregates per-run status — the Dagster/Prefect/Airflow backfill
// primitive.
//
// Backfill is idempotent: already-COMPLETED runs replay from storage in
// microseconds; already-FAILED runs re-execute (call inspector.ResetWorkflow
// first to clear them); IN_PROGRESS runs are reported as Pending and need a
// scanner / re-invoke to resume. Re-running Backfill over the same set is
// free for COMPLETED runs.
//
// Usage:
//
//	report, err := backfill.Backfill(ctx, runIDs, backfill.Options{
//	    Concurrency: 5,
//	}, func(ctx context.Context, runID string) (*pricesv1.FetchResponse, error) {
//	    return service.FetchPrices(ctx, connect.NewRequest(&pricesv1.FetchRequest{
//	        Symbol: "AAPL", RunId: runID,
//	    })).Msg, nil
//	})
package backfill

import (
	"context"
	"errors"
	"fmt"
	"sync"

	"connectrpc.com/connect"
	"github.com/jim-technologies/temporaless/core/go/workflow"
)

// Status of a single backfill entry.
type Status int

const (
	StatusUnspecified Status = iota
	StatusSucceeded
	StatusFailed
	// StatusPending means the workflow stayed IN_PROGRESS — a typed
	// pending error was returned, mapped to connect.CodeUnavailable, or
	// halt-on-error skipped this run_id.
	StatusPending
)

func (s Status) String() string {
	switch s {
	case StatusSucceeded:
		return "succeeded"
	case StatusFailed:
		return "failed"
	case StatusPending:
		return "pending"
	default:
		return "unspecified"
	}
}

// Entry is the result for one run_id.
type Entry[Resp any] struct {
	RunID  string
	Status Status
	Result Resp
	Err    error
}

// Report is the aggregate result of a Backfill call.
type Report[Resp any] struct {
	Entries []Entry[Resp]
}

// Succeeded returns the entries that completed cleanly.
func (r *Report[Resp]) Succeeded() []Entry[Resp] {
	return r.filter(StatusSucceeded)
}

// Failed returns the entries that ended with a non-pending error.
func (r *Report[Resp]) Failed() []Entry[Resp] {
	return r.filter(StatusFailed)
}

// Pending returns the entries that need to be re-driven (IN_PROGRESS).
func (r *Report[Resp]) Pending() []Entry[Resp] {
	return r.filter(StatusPending)
}

func (r *Report[Resp]) filter(status Status) []Entry[Resp] {
	out := make([]Entry[Resp], 0, len(r.Entries))
	for _, e := range r.Entries {
		if e.Status == status {
			out = append(out, e)
		}
	}
	return out
}

func (r *Report[Resp]) String() string {
	return fmt.Sprintf(
		"Report(succeeded=%d, failed=%d, pending=%d, total=%d)",
		len(r.Succeeded()),
		len(r.Failed()),
		len(r.Pending()),
		len(r.Entries),
	)
}

// Options configures a Backfill call.
type Options struct {
	// Concurrency is the maximum simultaneous in-flight invocations.
	// Zero means 1.
	Concurrency int
	// HaltOnError stops dispatching new run_ids after the first failure.
	// Already-running invocations finish; un-dispatched ones are reported
	// as Pending.
	HaltOnError bool
}

// Invoke is the user-supplied function the backfill loop calls per run_id.
type Invoke[Resp any] func(ctx context.Context, runID string) (Resp, error)

// Backfill runs invoke once per run_id with bounded concurrency and aggregates
// the results into a Report. Per-run errors are independent unless
// opts.HaltOnError is set.
//
// Workflow runs that stay IN_PROGRESS (TimerPendingError, EventPendingError,
// WorkflowDependencyPendingError, or their connect.CodeUnavailable mapped
// form) are reported as Pending — they aren't failures, they need a scanner
// to re-invoke them.
func Backfill[Resp any](
	ctx context.Context,
	runIDs []string,
	opts Options,
	invoke Invoke[Resp],
) (*Report[Resp], error) {
	if invoke == nil {
		return nil, errors.New("invoke is required")
	}
	concurrency := opts.Concurrency
	if concurrency == 0 {
		concurrency = 1
	}
	if concurrency < 1 {
		return nil, errors.New("concurrency must be >= 1")
	}

	entries := make([]Entry[Resp], len(runIDs))
	sem := make(chan struct{}, concurrency)
	halt := make(chan struct{})
	var haltOnce sync.Once
	closeHalt := func() { haltOnce.Do(func() { close(halt) }) }

	var wg sync.WaitGroup
	for i, runID := range runIDs {
		wg.Add(1)
		go func(idx int, rid string) {
			defer wg.Done()
			select {
			case <-halt:
				entries[idx] = Entry[Resp]{RunID: rid, Status: StatusPending}
				return
			case sem <- struct{}{}:
			}
			defer func() { <-sem }()

			select {
			case <-halt:
				entries[idx] = Entry[Resp]{RunID: rid, Status: StatusPending}
				return
			default:
			}

			result, err := invoke(ctx, rid)
			if err == nil {
				entries[idx] = Entry[Resp]{RunID: rid, Status: StatusSucceeded, Result: result}
				return
			}
			if isPendingError(err) {
				entries[idx] = Entry[Resp]{RunID: rid, Status: StatusPending, Err: err}
				return
			}
			if opts.HaltOnError {
				closeHalt()
			}
			entries[idx] = Entry[Resp]{RunID: rid, Status: StatusFailed, Err: err}
		}(i, runID)
	}
	wg.Wait()
	return &Report[Resp]{Entries: entries}, nil
}

func isPendingError(err error) bool {
	var timerPending *workflow.TimerPendingError
	if errors.As(err, &timerPending) {
		return true
	}
	var eventPending *workflow.EventPendingError
	if errors.As(err, &eventPending) {
		return true
	}
	var depPending *workflow.WorkflowDependencyPendingError
	if errors.As(err, &depPending) {
		return true
	}
	var connectErr *connect.Error
	if errors.As(err, &connectErr) && connectErr.Code() == connect.CodeUnavailable {
		return true
	}
	return false
}

File: core/py/src/temporaless/backfill.py
"""Backfill helper for running a workflow over many run_ids.

The Dagster/Prefect/Airflow backfill primitive: given a workflow and a set of
``run_ids`` (typically dates / partitions), dispatch each one with bounded
concurrency and report aggregate status.

Backfill is idempotent: already-COMPLETED runs replay from storage in
microseconds; already-FAILED runs re-execute (call ``inspector.reset_workflow``
first to clear them); IN_PROGRESS runs are reported as PENDING and need a
scanner / re-invoke to resume. Re-running ``backfill`` over the same set is
free for COMPLETED runs.

Example::

    from temporaless.backfill import backfill

    service = QuantService(store)

    async def invoke(run_id: str) -> StringValue:
        return await service.fetch_prices(StringValue(value=run_id))

    report = await backfill(
        invoke,
        run_ids=[f"2026-{m:02d}-{d:02d}" for m in (4, 5) for d in range(1, 31)],
        concurrency=5,
    )
    print(f"backfill: {report}")
    for entry in report.failed():
        print(f"  failed: {entry.run_id} → {entry.error}")
"""

from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable, Iterable
from dataclasses import dataclass, field
from enum import StrEnum

from temporaless.workflow import (
    ClaimBusyError,
    ConcurrencyBusyError,
    EventPendingError,
    TimerPendingError,
    WorkflowDependencyPendingError,
)


class BackfillStatus(StrEnum):
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    PENDING = "pending"


@dataclass
class BackfillEntry[T]:
    run_id: str
    status: BackfillStatus
    result: T | None = None
    error: BaseException | None = None


@dataclass
class BackfillReport[T]:
    entries: list[BackfillEntry[T]] = field(default_factory=list)

    def succeeded(self) -> list[BackfillEntry[T]]:
        return [e for e in self.entries if e.status == BackfillStatus.SUCCEEDED]

    def failed(self) -> list[BackfillEntry[T]]:
        return [e for e in self.entries if e.status == BackfillStatus.FAILED]

    def pending(self) -> list[BackfillEntry[T]]:
        return [e for e in self.entries if e.status == BackfillStatus.PENDING]

    def __str__(self) -> str:
        return (
            f"BackfillReport(succeeded={len(self.succeeded())}, "
            f"failed={len(self.failed())}, "
            f"pending={len(self.pending())}, "
            f"total={len(self.entries)})"
        )


async def backfill[T](
    invoke: Callable[[str], Awaitable[T]],
    run_ids: Iterable[str],
    *,
    concurrency: int = 1,
    halt_on_error: bool = False,
    pending_error: Callable[[BaseException], bool] | None = None,
) -> BackfillReport[T]:
    """Run a workflow over many run_ids with bounded concurrency.

    ``invoke`` is a partial-applied workflow caller: given a run_id, it returns
    the workflow result (typically by calling a transport adapter method that
    derives ``Options.run_id`` from its request). Per-run results are
    independent: a failure in one run_id doesn't affect others, unless
    ``halt_on_error=True``.

    Workflow runs that stay IN_PROGRESS (timer/event/dependency pending or a
    live claim holder), plus concurrency-cap contention, are PENDING. Wrapped
    local exceptions are recognized through their ``__cause__`` and
    ``__context__`` chains. A remote transport can pass ``pending_error`` to
    classify its status representation without coupling core to that
    transport.

    Args:
        invoke: callable mapping a run_id to a workflow result.
        run_ids: iterable of run_ids to dispatch.
        concurrency: maximum simultaneous in-flight invocations (default 1).
        halt_on_error: stop dispatching new run_ids after the first failure;
            already-running invocations finish; un-dispatched ones are reported
            as PENDING.
        pending_error: optional transport-specific predicate. It supplements
            the core typed-error classification; it must return ``True`` only
            for errors that are safe to retry as an IN_PROGRESS run.

    Returns:
        ``BackfillReport`` with one ``BackfillEntry`` per input run_id.
    """
    if concurrency < 1:
        raise ValueError("concurrency must be >= 1")
    if pending_error is not None and not callable(pending_error):
        raise TypeError("pending_error must be callable")

    semaphore = asyncio.Semaphore(concurrency)
    halt = asyncio.Event()

    async def run_one(run_id: str) -> BackfillEntry[T]:
        if halt.is_set():
            return BackfillEntry(run_id=run_id, status=BackfillStatus.PENDING)
        async with semaphore:
            if halt.is_set():
                return BackfillEntry(run_id=run_id, status=BackfillStatus.PENDING)
            try:
                result = await invoke(run_id)
                return BackfillEntry(run_id=run_id, status=BackfillStatus.SUCCEEDED, result=result)
            except Exception as exc:
                if _is_pending_error(exc) or (pending_error is not None and pending_error(exc)):
                    return BackfillEntry(run_id=run_id, status=BackfillStatus.PENDING, error=exc)
                if halt_on_error:
                    halt.set()
                return BackfillEntry(run_id=run_id, status=BackfillStatus.FAILED, error=exc)

    entries = await asyncio.gather(*(run_one(rid) for rid in run_ids))
    return BackfillReport(entries=list(entries))


def _is_pending_error(exc: BaseException) -> bool:
    pending_types = (
        TimerPendingError,
        EventPendingError,
        WorkflowDependencyPendingError,
        ClaimBusyError,
        ConcurrencyBusyError,
    )
    stack = [exc]
    seen: set[int] = set()
    while stack:
        current = stack.pop()
        identity = id(current)
        if identity in seen:
            continue
        seen.add(identity)
        if isinstance(current, pending_types):
            return True
        if current.__cause__ is not None:
            stack.append(current.__cause__)
        if current.__context__ is not None:
            stack.append(current.__context__)
    return False
